# Reporte: Predicción del Laberinto - Cadena de Markov

Este reporte analiza el comportamiento de un ratón en un laberinto de 6 casillas utilizando Cadenas de Markov.

## 1. Datos del problema

- **Laberinto:** 6 casillas numeradas del 1 al 6.
- **Probabilidades iniciales:** La probabilidad de introducir al ratón en las casillas 1, 2 y 3 es el doble que en las casillas 4, 5 y 6.
- **Regla de movimiento:** El ratón se mueve a cualquier casilla contigua o se queda en la misma con la misma probabilidad.
- **Estructura de transición (Matriz P):**

$$P = \begin{pmatrix}
p_{11} & p_{12} & p_{13} & 0 & 0 & 0 \\
p_{21} & p_{22} & 0 & p_{24} & 0 & 0 \\
p_{31} & 0 & p_{33} & p_{34} & p_{35} & 0 \\
0 & p_{42} & p_{43} & p_{44} & 0 & p_{46} \\
0 & 0 & p_{53} & 0 & p_{55} & p_{56} \\
0 & 0 & 0 & p_{64} & p_{65} & p_{66}
\end{pmatrix}$$

  Se define por la conectividad entre casillas:
  - Casilla 1 conecta con: 1, 2, 3 (3 opciones)
  - Casilla 2 conecta con: 1, 2, 4 (3 opciones)
  - Casilla 3 conecta con: 1, 3, 4, 5 (4 opciones)
  - Casilla 4 conecta con: 2, 3, 4, 6 (4 opciones)
  - Casilla 5 conecta con: 3, 5, 6 (3 opciones)
  - Casilla 6 conecta con: 4, 5, 6 (3 opciones)

## 2. Abstracción a una cadena de Markov

Definimos el proceso estocástico $\{X_n, n \geq 0\}$ donde $X_n$ representa la casilla en la que se encuentra el ratón en el instante de tiempo $n$. El espacio de estados es $S = \{1, 2, 3, 4, 5, 6\}$. Dado que la probabilidad del siguiente estado depende únicamente del estado actual, el proceso cumple con la propiedad de Markov.

## 3. Vector de probabilidades iniciales y matriz de transición

### Vector de probabilidades iniciales ($\pi_0$)
Sean $p$ la probabilidad para las casillas 4, 5 y 6, y $2p$ para las casillas 1, 2 y 3.
La suma de probabilidades debe ser 1:
$2p + 2p + 2p + p + p + p = 1 \implies 9p = 1 \implies p = 1/9$.

Por lo tanto:
$\pi_0 = [\frac{2}{9}, \frac{2}{9}, \frac{2}{9}, \frac{1}{9}, \frac{1}{9}, \frac{1}{9}]$

### Matriz de transición (P)
Basándonos en la conectividad y la regla de equiprobabilidad:
- $p_{11}=p_{12}=p_{13} = 1/3$
- $p_{21}=p_{22}=p_{24} = 1/3$
- $p_{31}=p_{33}=p_{34}=p_{35} = 1/4$
- $p_{42}=p_{43}=p_{44}=p_{46} = 1/4$
- $p_{53}=p_{55}=p_{56} = 1/3$
- $p_{64}=p_{65}=p_{66} = 1/3$


In [12]:
import numpy as np
import pandas as pd

# Definición del vector de probabilidades iniciales
pi_0 = np.array([2/9, 2/9, 2/9, 1/9, 1/9, 1/9])

P = np.array([
    [1/3, 1/3, 1/3, 0,   0,   0],
    [1/3, 1/3, 0,   1/3, 0,   0],
    [1/4, 0,   1/4, 1/4, 1/4, 0],
    [0,   1/4, 1/4, 1/4, 0,   1/4],
    [0,   0,   1/3, 0,   1/3, 1/3],
    [0,   0,   0,   1/3, 1/3, 1/3]
])

print("Vector pi_0:")
print(pi_0)
print("\nMatriz de transición P:")
print(P)


Vector pi_0:
[0.22222222 0.22222222 0.22222222 0.11111111 0.11111111 0.11111111]

Matriz de transición P:
[[0.33333333 0.33333333 0.33333333 0.         0.         0.        ]
 [0.33333333 0.33333333 0.         0.33333333 0.         0.        ]
 [0.25       0.         0.25       0.25       0.25       0.        ]
 [0.         0.25       0.25       0.25       0.         0.25      ]
 [0.         0.         0.33333333 0.         0.33333333 0.33333333]
 [0.         0.         0.         0.33333333 0.33333333 0.33333333]]


## 4. Probabilidades de transición correspondientes a la matriz

A continuación se muestra la matriz P con sus valores calculados:


In [13]:
df_P = pd.DataFrame(P, columns=[1,2,3,4,5,6], index=[1,2,3,4,5,6])
df_P


,1,2,3,4,5,6
1,0.333333,0.333333,0.333333,0.000000,0.000000,0.000000
2,0.333333,0.333333,0.000000,0.333333,0.000000,0.000000
3,0.250000,0.000000,0.250000,0.250000,0.250000,0.000000
4,0.000000,0.250000,0.250000,0.250000,0.000000,0.250000
5,0.000000,0.000000,0.333333,0.000000,0.333333,0.333333
6,0.000000,0.000000,0.000000,0.333333,0.333333,0.333333


## 5. Calcule la probabilidad de estar en cierta casilla en el instante de tiempo 2 y 3

Para calcular la distribución de probabilidad en el instante $n$, $\pi_n$, utilizamos la fórmula de la cadena de Markov:
$$\pi_n = \pi_0 P^n$$

### Cálculos para n=2 y n=3:


In [14]:
# Probabilidades en t=2
pi_2 = pi_0 @ np.linalg.matrix_power(P, 2)

# Probabilidades en t=3
pi_3 = pi_0 @ np.linalg.matrix_power(P, 3)

print("Vector de probabilidades en t=2 (pi_2):")
for i, prob in enumerate(pi_2, 1):
    print(f"Casilla {i}: {prob:.4f}")

print("\nVector de probabilidades en t=3 (pi_3):")
for i, prob in enumerate(pi_3, 1):
    print(f"Casilla {i}: {prob:.4f}")


Vector de probabilidades en t=2 (pi_2):
Casilla 1: 0.1752
Casilla 2: 0.1752
Casilla 3: 0.2083
Casilla 4: 0.1898
Casilla 5: 0.1258
Casilla 6: 0.1258

Vector de probabilidades en t=3 (pi_3):
Casilla 1: 0.1689
Casilla 2: 0.1642
Casilla 3: 0.1998
Casilla 4: 0.1998
Casilla 5: 0.1359
Casilla 6: 0.1313


### Resultados finales

En el instante de tiempo **t=2**, las probabilidades son aproximadamente:
- Casilla 1: 0.1752
- Casilla 2: 0.1752
- Casilla 3: 0.2083
- Casilla 4: 0.1898
- Casilla 5: 0.1258
- Casilla 6: 0.1258

En el instante de tiempo **t=3**, las probabilidades son aproximadamente:
- Casilla 1: 0.1689
- Casilla 2: 0.1642
- Casilla 3: 0.1998
- Casilla 4: 0.1998
- Casilla 5: 0.1359
- Casilla 6: 0.1313
